### ISSUE TITLE
[dynamo] HF transformers layerdrop idiom: data-dep branch on torch.rand([]) skips train-mode compile

### ISSUE THREAD

**Issue by penguinwu** [penguinwu/oss-model-graph-break-corpus#77]
## Summary

HF transformers' standard layerdrop idiom triggers a data-dependent graph break in train mode: `dropout_probability = torch.rand([])` followed by `if dropout_probability < self.layerdrop: continue` is not traceable by Dynamo.

**Why this matters:** the same layerdrop idiom appears in roughly 89 HF transformer model families — Bart, MBart, Marian, M2M100, Pegasus, Blenderbot, MusicGen, OPT, BigBirdPegasus, NllbMoe, SeamlessM4T, BioGpt, RagModel, AudioFlamingo, and others (full list in the affected-models table below). Train-mode `torch.compile` is blocked on every model that uses the idiom; eval mode is unaffected because layerdrop is bypassed when `self.training == False`. Resolving this pattern would unblock train-mode compile across a large fraction of HF seq2seq, audio, and multimodal models.

## Reproduction

```python
import torch
from transformers import BartForConditionalGeneration, BartConfig

torch.manual_seed(42)

# Tiny config; layerdrop > 0 to trigger the train-mode idiom.
config = BartConfig(
    vocab_size=100, d_model=32,
    encoder_layers=2, decoder_layers=2,
    encoder_attention_heads=2, decoder_attention_heads=2,
    encoder_ffn_dim=64, decoder_ffn_dim=64,
    encoder_layerdrop=0.5, decoder_layerdrop=0.5,
    max_position_embeddings=64,
)
model = BartForConditionalGeneration(config).train()

input_ids = torch.zeros(1, 8, dtype=torch.long)
decoder_input_ids = torch.zeros(1, 4, dtype=torch.long)

torch.compile(model, fullgraph=True)(
    input_ids=input_ids, decoder_input_ids=decoder_input_ids,
)
```

**Run:** `python repro.py` (torch nightly 2026-05-02 + transformers 5.6.2; captured 2026-05-08).

## Observed output

```
torch._dynamo.exc.Unsupported: Data-dependent branching
  Explanation: Detected data-dependent branching (e.g. `if my_tensor.sum() > 0:`). Dynamo does not support tracing dynamic control flow.

      The branch condition involves a tensor computed as follows:
        # File "transformers/models/bart/modeling_bart.py", line 536, in forward, code: dropout_probability = torch.rand([])
        dropout_probability = rand([])

        # File "transformers/models/bart/modeling_bart.py", line 537, in forward, code: if dropout_probability < self.layerdrop:  # skip the layer
        lt = lt(dropout_probability, 0.5)

  Hint: For the common pattern `if tensor_cond: x = transform(x)` (e.g. clamping inf/nan values), consider making the code branchless by always applying the transform. Operations like torch.clamp, torch.nan_to_num, and torch.where are typically no-ops on well-behaved inputs and compile without graph breaks.
  Hint: This graph break is fundamental - it is unlikely that Dynamo will ever be able to trace through your code. Consider finding a workaround.
  Hint: Use `torch.cond` to express dynamic control flow.

  Developer debug context: attempted to jump with TensorVariable()

 For more details about this graph break, please visit: https://meta-pytorch.github.io/compile-graph-break-site/gb/gb0170.html
```

## Pattern

```python
# transformers layerdrop idiom (encoder/decoder forward, train mode):
dropout_probability = torch.rand([])
if dropout_probability < self.layerdrop:
    continue
```

The same idiom is copy-pasted across HF model families wherever layerdrop is implemented. The line numbers shift across transformers versions (e.g. line 536-537 in transformers 5.6.2 Bart) but the pattern is identical.

## Source

- Source pattern (transformers 5.6.2): `transformers/models/bart/modeling_bart.py:536-537` (encoder layerdrop). Decoder layerdrop is the same idiom further down the same file.
- Umbrella: this issue is a sub-pattern of #54 (data-dependent branching, broader).

## Environment

- torch: `2.13.0.dev20260502+cu126`
- transformers: `5.6.2`
- diffusers: `0.38.0`

## Affected models

Compiled from a sweep of HF transformer models with layerdrop > 0 in train mode. Per-model break counts are approximate (collected against transformers 5.6.2 / torch nightly 2026-05-02; counts will shift across stack versions but the affected-set is stable wherever layerdrop is implemented).

<details>
<summary>89-model affected-list (click to expand)</summary>

| Model | Modes | Breaks |
|-------|-------|--------|
| AudioFlamingo3ForConditionalGeneration | train | 7 |
| AutoformerModel | train | 5 |
| BartForCausalLM | train | 5 |
| BartForConditionalGeneration | train | 9 |
| BartModel | train | 7 |
| BigBirdPegasusForCausalLM | train | 5 |
| BigBirdPegasusForConditionalGeneration | train | 3 |
| BigBirdPegasusModel | train | 3 |
| BioGptForCausalLM | train | 5 |
| BioGptModel | train | 3 |
| BlenderbotForCausalLM | train | 5 |
| BlenderbotForConditionalGeneration | train | 9 |
| BlenderbotModel | train | 7 |
| BlenderbotSmallForCausalLM | train | 5 |
| BlenderbotSmallForConditionalGeneration | train | 9 |
| BlenderbotSmallModel | train | 7 |
| Blip2ForConditionalGeneration | train | 7 |
| Blip2Model | train | 7 |
| CohereAsrForConditionalGeneration | train | 8 |
| CohereAsrModel | train | 6 |
| Data2VecAudioModel | train | 2 |
| FSMTForConditionalGeneration | train | 5 |
| FSMTModel | train | 4 |
| FlaubertModel | train | 1 |
| FlaubertWithLMHeadModel | train | 1 |
| HubertModel | train | 2 |
| InformerModel | train | 6 |
| InstructBlipForConditionalGeneration | train | 7 |
| InstructBlipModel | train | 7 |
| InstructBlipVideoForConditionalGeneration | train | 7 |
| InstructBlipVideoModel | train | 7 |
| Kosmos2ForConditionalGeneration | train | 7 |
| Kosmos2Model | train | 7 |
| LEDForConditionalGeneration | train | 3 |
| LEDModel | train | 3 |
| M2M100ForConditionalGeneration | train | 7 |
| M2M100Model | train | 6 |
| MBartForCausalLM | train | 5 |
| MBartForConditionalGeneration | train | 4 |
| MBartModel | train | 4 |
| MarianForCausalLM | train | 5 |
| MarianMTModel | train | 9 |
| MarianModel | train | 7 |
| MusicgenForCausalLM | train | 5 |
| MusicgenForConditionalGeneration | train | 7 |
| MusicgenMelodyForCausalLM | train | 5 |
| MusicgenMelodyForConditionalGeneration | train | 6 |
| MusicgenMelodyModel | train | 4 |
| MusicgenModel | train | 4 |
| MvpForCausalLM | train | 2 |
| MvpForConditionalGeneration | train | 5 |
| MvpModel | train | 4 |
| NllbMoeForConditionalGeneration | train | 9 |
| NllbMoeModel | train | 7 |
| OPTForCausalLM | train | 5 |
| OPTModel | train | 5 |
| PLBartForCausalLM | train | 5 |
| PLBartForConditionalGeneration | train | 7 |
| PLBartModel | train | 6 |
| PegasusForCausalLM | train | 5 |
| PegasusForConditionalGeneration | train | 9 |
| PegasusModel | train | 7 |
| PegasusXForConditionalGeneration | train | 9 |
| PegasusXModel | train | 7 |
| Qwen2AudioForConditionalGeneration | train | 5 |
| RagModel | train | 10 |
| SEWModel | train | 2 |
| SeamlessM4TModel | train | 4 |
| SeamlessM4TTextToUnitForConditionalGeneration | train | 5 |
| SeamlessM4TTextToUnitModel | train | 4 |
| SeamlessM4Tv2Model | train | 4 |
| Speech2TextForConditionalGeneration | train | 7 |
| Speech2TextModel | train | 6 |
| SpeechEncoderDecoderModel | train | 2 |
| SpeechT5Model | train | 6 |
| TimeSeriesTransformerModel | train | 6 |
| TrOCRForCausalLM | train | 2 |
| UniSpeechModel | train | 2 |
| UniSpeechSatModel | train | 2 |
| Wav2Vec2BertModel | train | 2 |
| Wav2Vec2ConformerModel | train | 2 |
| Wav2Vec2Model | train | 2 |
| WavLMModel | train | 2 |
| WhisperForCausalLM | train | 5 |
| WhisperForConditionalGeneration | train | 9 |
| WhisperModel | train | 7 |
| XGLMForCausalLM | train | 4 |
| XGLMModel | train | 3 |
| XcodecModel | train | 2 |

## Impact

Fixing this pattern eliminates ~466 train-mode breaks across 89 models. Per-model fullgraph-win count requires re-sweep.

## Related Issues

- #54 (umbrella: data-dependent branching, all sub-patterns)

## Tested On

- PyTorch 2.10.0a0+gitf32431e (with PR #181552)
- Transformers 5.6.2 (HEAD as of 2026-04-26)
- Sweep date: 2026-04-28
- Sweep dir: `sweep_results/experiments/animesh-fullgraph-2026-04-28`

</details>

**Comment by penguinwu** [comment]
## Substantial progress on 2026-05-09 nightly sweep — 75% of affected models now compile fullgraph

Of the 89 models tracked in this issue, **67 now compile fullgraph** (status flipped from `graph_break` → `full_graph` between 2026-05-03 and 2026-05-09 sweeps). 22 still hit the layerdrop idiom break.

The remaining 22 models are likely those whose layerdrop call site differs from the dominant pattern (different transformer family, different control-flow shape). Leaving this issue OPEN to track the remainder; not auto-closing since partial fix means the underlying gap may still affect some idioms.

Sweep results: `sweep_results/nightly/2026-05-09/identify_results.json`.

<sub>Status update via 2026-05-09 nightly sweep close-stale check (75% reduction). Auto-close threshold: 100%.</sub>
